cascadenet/network/layers/kspace_pytorch.py

```diff
@@ -0,0 +1,440 @@
+            acc[:,:, :-d] = acc[:,:, :-d] + x[:,:, d:]
+            mask_acc[:,:, d:] =  mask_acc[:,:, d:] + mask[:,:, :-d]
+            mask_acc[:,:, :-d] = mask_acc[:,:, :-d] + mask[:,:, d:]
+
+            if not self.clipped:
+                # add the part which wraps around
+                acc[:,:, :d] = acc[:,:, :d] + x[:,:, -d:]
+                acc[:,:, -d:] = acc[:,:, -d:] + x[:,:, :d]
+                mask_acc[:,:, :d] = mask_acc[:,:, :d] + mask[:,:, -d:]
+                mask_acc[:,:, -d:] = mask_acc[:,:, -d:] + mask[:,:, :d]
+
+
+            if d in self.frame_dist:
+                # TODO check if it is acc.clone() or acc.data.clone()
+                accs.append(acc.clone())
+                mask_accs.append(mask_acc.clone())
+
+        # divide unbiasedly
+        for i, t in enumerate(self.n_samples):
+            if self.divide_by_n:
+                c = t
+            else:
+                c = 1
+
+            accs[i] = accs[i] / mask_accs[i].clamp(min=c)
+            accs[i] = (1 - mask) * accs[i] + x
+
+
+            # for conveniently concatenating
+            accs[i] = accs[i].unsqueeze(2)
+
+        accs.insert(0, x.unsqueeze(2))
+
+        return torch.cat(accs, dim=2)
+
+
+class AveragingInKspace(nn.Module):
+    '''
+    Average-in-k-space layer
+
+    First transforms the representation in Fourier domain,
+    then performs averaging along temporal axis, then transforms back to image
+    domain. Works only for 5D tensor (see parameter descriptions).
+
+
+    Parameters
+    -----------------------------
+    incomings: two 5d tensors, [kspace_data, mask], each of shape (n, 2, nx, ny, nt)
+
+    data_shape: shape of the incoming tensors: (n, 2, nx, ny, nt) (This is for convenience)
+
+    frame_dist: a list of distances of neighbours to sample for each averaging channel
+        if frame_dist=[1], samples from [-1, 1] for each temporal frames
+        if frame_dist=[3, 5], samples from [-3,-2,...,0,1,...,3] for one,
+                                           [-5,-4,...,0,1,...,5] for the second one
+
+    divide_by_n: bool - Decides how averaging will be done.
+        True => divide by number of neighbours (=#2*frame_dist+1)
+        False => divide by number of nonzero contributions
+
+    clipped: bool - By default the layer assumes periodic boundary condition along temporal axis.
+        True => Averaging will be clipped at the boundary, no circular references.
+        False => Averages with circular referencing (i.e. at t=0, gets contribution from t=nt-1, so on).
+
+    Returns
+    ------------------------------
+    output: 5d tensor, missing lines of k-space are filled using neighbouring frames.
+            shape becomes (n, (len(frame_dist))* 2, nx, ny, nt)
+    '''
+
+    def __init__(self, frame_dist, divide_by_n=False, clipped=True, norm='ortho'):
+        super(AveragingInKspace, self).__init__()
+        self.normalized = norm == 'ortho'
+        # self.fft2_fun = Fft2d(norm)
```

Comment:
Delete these comments. Can you also make sure we don't have these "commented out" ones throught? Thanks!